[SelectionDAG][x86] Ensure vector reduction optimization

sutajo opened the pull request:
Fixes #144227

sutajo commented:
I agree that it makes little sense to expand the reductions and then later recombine them. 
Alright, I'll look into lowering some reduction OPs in the x86 backend.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -25409,6 +25423,94 @@ static SDValue LowerEXTEND_VECTOR_INREG(SDValue Op,
+    case ISD::VECREDUCE_SMAX:
+      BinOp = ISD::SMAX;
+      break;
+  }
```
replace this with `ISD::NodeType BinOp = ISD::getVecReduceBaseOpcode(Op.getOpcode())` - you can then assert that BinOp is a min/max opcode if you want.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -1435,6 +1435,20 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
+  // Vector min/max reductions
+  if (Subtarget.hasSSE41())
+  {
+    for (MVT VT : MVT::vector_valuetypes()) {
```
SSE should only be declaring v816/v16i8 custom if we can.

sutajo replied:
Now I see what you mean. I added checks here to only do lowering for vector types that support PHMINPOSUW.

RKSimon reviewed (commented):
please can you clang-format your changes?

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -46192,60 +46299,8 @@ static SDValue combineMinMaxReduction(SDNode *Extract, SelectionDAG &DAG,
-
-  return DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, ExtractVT, MinPos,
-                     DAG.getVectorIdxConstant(0, DL));
+  return createMinMaxReduction(Src, ExtractVT, SDLoc(Extract),
```
Is this still required?

RKSimon replied:
We have some CodeGen/X86 tests that match the raw shuffle chain pattern instead of the reduction intrinsic that fail without this - its probably another case of #143088 that we need to improve the middle-end to ensure those patterns get folded into reduction intrinsics so the backend doesn't need to handle it

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -47255,6 +47321,36 @@ static SDValue combineExtractVectorElt(SDNode *N, SelectionDAG &DAG,
   return SDValue();
 }
 
+static SDValue combineExtractVectorEltAndOperand(SDNode* N, SelectionDAG& DAG,
```
Is the combine independent of moving code to the lowering? Better to split them if it is.

sutajo replied:
They are independent, but in the end they serve the same purpose. Should I open a separate PR for the lowering?

phoebewang replied:
Yes. That would be good to show which code affect the test change.

RKSimon reviewed (changes requested):
please can you remove the cmake and .gitignore diffs

sutajo commented:
Oops 😅

sutajo commented:
I reverted the custom lowering here and moved it to [x86-custom-lower-vecreduce](https://github.com/sutajo/llvm-project/tree/x86-custom-lower-vecreduce)

It is still WIP, I plan to add also xor/or/and and after that I will create a PR.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -48066,6 +48077,36 @@ static SDValue combineExtractVectorElt(SDNode *N, SelectionDAG &DAG,
+  auto Op = combineExtractVectorElt(N, DAG, DCI, Subtarget, TransformedBinOpReduction);
+
+  if (TransformedBinOpReduction)
+  {
```
clang format this section

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -48066,6 +48077,36 @@ static SDValue combineExtractVectorElt(SDNode *N, SelectionDAG &DAG,
+
+    auto NewV = DAG.getNode(ISD::SCALAR_TO_VECTOR, SDLoc(N), OldV->getValueType(0), Op);
+
+    auto NV = DCI.CombineTo(N, Op);
```
(style) don't use auto

RKSimon reviewed (changes requested):
apologies @sutajo this fell off my radar - I've added the tests at #193484 (replacing your optimize-reduction.ll) - please can you merge against trunk and regenerate to show the codegen diffs?

sutajo committed:
[x86] Fully eliminate shuffle pyramid when combining binop reductions (#144231)

sutajo commented:
@RKSimon, I rebased my changes ontop of main and regenerated the test that you added in https://github.com/llvm/llvm-project/pull/193484 so that you can see the diff.

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -47873,26 +47892,29 @@ static SDValue combineExtractVectorElt(SDNode *N, SelectionDAG &DAG,
+      (BinOpReduction = combinePredicateReduction(N, DAG, Subtarget)) ||
+      ((Result = combineMinMaxReduction(N, DAG, Subtarget)), BinOpReduction) ||
+      ((Result = combineArithReduction(N, DAG, Subtarget)), BinOpReduction)) {
+    SDValue ExtractEltOperand = N->getOperand(0);
```
Move this inside `if (!IsPartialReduction)`

sutajo replied:
Not possible, because `N` gets deleted by `DCI.CombineTo` and I can’t access the operand.

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -47873,26 +47892,29 @@ static SDValue combineExtractVectorElt(SDNode *N, SelectionDAG &DAG,
+      DCI.CombineTo(ExtractEltOperand.getNode(), V);
+    }
+
+    return SDValue(N, 0); // Return N so it doesn't get rechecked!
```
Can we return BinOpReduction instead?

sutajo replied:
Order is important here, `N` needs to be swapped first, otherwise `DCI.CombineTo(ExtractEltOperand.getNode(), V)` could recursively delete `N`. Which means we can't reutrn `BinOpReduction`, because that would cause `N` to be recombined after `ExtractEltOperand`.

phoebewang replied:
Can it be:
```
if (IsPartialReduction)
  return BinOpReduction;

...
```

sutajo replied:
Yes, that works. Replaced it.

phoebewang reviewed (approved):
LGTM.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -47156,6 +47162,9 @@ static SDValue combineVPDPBUSDPattern(SDNode *Extract, SelectionDAG &DAG,
 
+// Check whether this extract is the root of a sum of absolute differences
+// pattern. This has to be done here because we really want it to happen
+// pre-legalizatio.
```
```suggestion
// pre-legalization.
```

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -47873,26 +47892,29 @@ static SDValue combineExtractVectorElt(SDNode *N, SelectionDAG &DAG,
+      (BinOpReduction = combineVPDPBUSDPattern(N, DAG, Subtarget)) ||
+      (BinOpReduction = combinePredicateReduction(N, DAG, Subtarget)) ||
+      ((Result = combineMinMaxReduction(N, DAG, Subtarget)), BinOpReduction) ||
+      ((Result = combineArithReduction(N, DAG, Subtarget)), BinOpReduction)) {
```
I don't understand why you've pulled in all the other combines when #144227 is just about minmax/arith reductions

ts-numfum replied:
The issue is not only related to minmax/arith reductions. I discovered the issue with min reductions, that is why I used that as an example.
The problem can manifest itself with any binop reduction.

RKSimon commented:
I'm not certain I like the direction this is going - I'd be much happier if x86 finally started work on supporting the ISD::VECREDUCE_* nodes instead of adding further kludges.
